wpilibj/src/main/java/org/wpilib/framework/OpModeRobot.java

```diff
@@ -489,6 +493,40 @@ public void driverStationConnected() {}
    */
   public void nonePeriodic() {}
 
+  /**
+   * Function called periodically by {@link PeriodicOpMode} following the OpMode-specific
+   * {@link PeriodicOpMode#periodic} function. This is not called by {@link LinearOpMode}.
+   */
+  public void robotPeriodic() {}
+
+  /**
+   * Function called periodically during simulation by {@link PeriodicOpMode} following {@link OpModeRobot#robotPeriodic}.
+   * This is not called by {@link LinearOpMode}.
+   */
+  public void simulationPeriodic() {}
+
+  /**
+   * Internal periodic function. This is responsible for calling {@link OpModeRobot#robotPeriodic} and
+   * {@link OpModeRobot#simulationPeriodic} (during simulation).
+   * {@link PeriodicOpMode} automatically calls this every loop, but {@link LinearOpMode} does not.
+   *
+   * @param watchdog watchdog instance, typically passed in from the calling {@link PeriodicOpMode}.
+   */
+  public final void internalRobotPeriodic(Watchdog watchdog) {
```

Comment:
Maybe I missed something while reading through chat history, but why make a separate internal method here instead of inlining it into `PeriodicOpMode.loopFunc()`?